tests/unit/test_param_expand.py

```diff
@@ -344,3 +347,122 @@ def test_parameter_graph_mixing_offset_and_conditional(self):
+    ):# -> None:
+        """Ease of reading wrapper for pytest.param
+
+        Args:
+            expect:
+                Output of expand_parent_params()
+            raw_str:
+                The parent_params input string.
+            parameter_values
+        """
+        self.xfail = xfail
+        self.raises = raises
+        self.expect = expect
+        self.raw_str = raw_str
+        self.parameter_values = parameter_values
+        self.templates = templates
+        self.parameters = ((parameter_values, templates))
+        self.name_expander = NameExpander(self.parameters)
+        self.id = 'raises:' + id if raises else id
+
+    def get(self):
+        return param(self, id=self.id)
+
+
+@pytest.mark.parametrize(
+    "param",
+    (
+        myParam(
+            expect=(None, 'no_params_here'),
+            raw_str='no_params_here',
+            id='basic'
+        ).get(),
+        myParam(
+            expect=({'bar': 1}, 'bar1'),
+            raw_str='<bar>',
+            parameter_values={'bar': 1},
+            templates={'bar': 'bar%(bar)s'},
+            id='one-valid-param'
+        ).get(),
+        myParam(
+            expect=({'bar': 1}, 'foo_bar1_baz'),
+            raw_str='foo<bar>baz',
+            parameter_values={'bar': 1},
+            templates={'bar': '_bar%(bar)s_'},
+            id='one-valid-param'
+        ).get(),
+        myParam(
+            raw_str='foo<bar>baz',
+            parameter_values={'qux': 2},
+            templates={'bar': '_bar%(bar)s_'},
+            raises=(ParamExpandError, 'parameter \'bar\' undefined'),
+            id='one-invalid-param'
+        ).get(),
+        myParam(
+            expect=({'bar': 1, 'baz': 42}, 'foo_bar1_baz42'),
+            raw_str='foo<bar, baz>',
+            parameter_values={'bar': 1, 'baz': 42},
+            templates={'bar': '_bar%(bar)s', 'baz': '_baz%(baz)s'},
+            id='two-valid-param'
+        ).get(),
+        myParam(
+            expect=({'bar': 1, 'baz': 42}, 'foo_bar1qux_baz42'),
+            raw_str='foo<bar>qux<baz>',
+            parameter_values={'bar': 1, 'baz': 42},
+            templates={'bar': '_bar%(bar)s', 'baz': '_baz%(baz)s'},
+            id='two-valid-param-sep-brackets',
+        ).get(),
+        myParam(
+            raw_str='foo<bar-1>baz',
+            raises=(ParamExpandError, '^parameter offsets illegal here'),
+            id='offsets-illegal'
+        ).get(),
+        myParam(
+            expect=({'bar': 1}, 'foo_bar1_baz'),
+            raw_str='foo<bar=1>baz',
+            parameter_values={'bar': [1, 2]},
+            templates={'bar': '_bar%(bar)s_'},
+            id='value-set'
+        ).get(),
+        myParam(
+            raw_str='foo<bar=3>baz',
+            parameter_values={'bar': [1, 2]},
+            raises=(ParamExpandError, '^illegal'),
+            id='illegal-value'
+        ).get(),
+        myParam(
+            expect=({'bar': 1}, 'foo_bar1_baz'),
+            raw_str='foo<bar=3>baz',
```

Comment:
Would be worth testing param indexing e.g. `<foo - 1, bar + 2>`

Reply:
See line 429?